cpp/include/cudf/lists/detail/scatter.cuh

```diff
@@ -0,0 +1,727 @@
+    cudf::lists_column_view const& target_lists_column_view,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream) const
+  {
+    auto source_column_device_view =
+      column_device_view::create(source_lists_column_view.parent(), stream);
+    auto target_column_device_view =
+      column_device_view::create(target_lists_column_view.parent(), stream);
+    auto source_lists = cudf::detail::lists_column_device_view(*source_column_device_view);
+    auto target_lists = cudf::detail::lists_column_device_view(*target_column_device_view);
+
+    int32_t num_child_rows{get_num_child_rows(list_offsets, stream)};
+
+    auto string_views = rmm::device_vector<string_view>(num_child_rows);
+
+    auto populate_string_views = [d_scattered_lists =
+                                    list_vector.data().get(),  // unbound_list_view*
+                                  d_list_offsets = list_offsets.template data<int32_t>(),
+                                  d_string_views = string_views.data().get(),
+                                  source_lists,
+                                  target_lists] __device__(auto const& row_index) {
+      auto unbound_list_view    = d_scattered_lists[row_index];
+      auto actual_list_row      = unbound_list_view.bind_to_column(source_lists, target_lists);
+      auto lists_column         = actual_list_row.get_column();
+      auto lists_offsets_column = lists_column.offsets();
+      auto child_strings_column = lists_column.child();
+      auto string_offsets_column =
+        child_strings_column.child(cudf::strings_column_view::offsets_column_index);
+      auto string_chars_column =
+        child_strings_column.child(cudf::strings_column_view::chars_column_index);
+
+      auto output_start_offset =
+        d_list_offsets[row_index];  // Offset in `string_views` at which string_views are
+                                    // to be written for this list row_index.
+      auto input_list_start =
+        lists_offsets_column.template element<int32_t>(unbound_list_view.row_index());
+
+      thrust::for_each_n(
+        thrust::seq,
+        thrust::make_counting_iterator<size_type>(0),
+        actual_list_row.size(),
+        [output_start_offset,
+         d_string_views,
+         input_list_start,
+         d_string_offsets = string_offsets_column.template data<int32_t>(),
+         d_string_chars =
+           string_chars_column.template data<char>()] __device__(auto const& string_idx) {
+          // auto string_offset     = output_start_offset + string_idx;
+          auto string_start_idx = d_string_offsets[input_list_start + string_idx];
+          auto string_end_idx   = d_string_offsets[input_list_start + string_idx + 1];
+
+          d_string_views[output_start_offset + string_idx] =
+            string_view{d_string_chars + string_start_idx, string_end_idx - string_start_idx};
+        });
+    };
+
+    thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
```

Comment:
I think the same optimization should be possible here to collapse and increase parallelism.